storage-proofs/porep/src/stacked/vanilla/proof.rs

```diff
@@ -480,66 +725,103 @@ impl<'a, Tree: 'static + MerkleTreeTrait, G: 'static + Hasher> StackedDrg<'a, Tr
         let labels = LabelsCache::<Tree>::new(&label_configs)?;
         let configs = split_config(tree_c_config.clone(), tree_count)?;
 
-        // Build the tree for CommC
-        let tree_c = measure_op(GenerateTreeC, || {
-            info!("Building column hashes");
-
-            let mut trees = Vec::with_capacity(tree_count);
-            for (i, config) in configs.iter().enumerate() {
-                let mut hashes: Vec<<Tree::Hasher as Hasher>::Domain> =
-                    vec![<Tree::Hasher as Hasher>::Domain::default(); nodes_count];
-
-                rayon::scope(|s| {
-                    // spawn n = num_cpus * 2 threads
-                    let n = num_cpus::get() * 2;
-
-                    // only split if we have at least two elements per thread
-                    let num_chunks = if n > nodes_count * 2 { 1 } else { n };
-
-                    // chunk into n chunks
-                    let chunk_size = (nodes_count as f64 / num_chunks as f64).ceil() as usize;
-
-                    // calculate all n chunks in parallel
-                    for (chunk, hashes_chunk) in hashes.chunks_mut(chunk_size).enumerate() {
-                        let labels = &labels;
-
-                        s.spawn(move |_| {
-                            for (j, hash) in hashes_chunk.iter_mut().enumerate() {
-                                let data: Vec<_> = (1..=layers)
-                                    .map(|layer| {
-                                        let store = labels.labels_for_layer(layer);
-                                        let el: <Tree::Hasher as Hasher>::Domain = store
-                                            .read_at((i * nodes_count) + j + chunk * chunk_size)
-                                            .unwrap();
-                                        el.into()
-                                    })
-                                    .collect();
-
-                                *hash = hash_single_column(&data).into();
-                            }
-                        });
-                    }
-                });
-
-                info!("building base tree_c {}/{}", i + 1, tree_count);
-                trees.push(DiskTree::<
-                    Tree::Hasher,
-                    Tree::Arity,
-                    typenum::U0,
-                    typenum::U0,
-                >::from_par_iter_with_config(
-                    hashes.into_par_iter(), config.clone()
-                )?);
+        let tree_c_root = match layers {
```

Comment:
@dignifiedquire If you get a chance, I could use some help making this part manageable.  The double layer matching is cumbersome at best ...